cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/kafkaassigner/KafkaAssignerEvenRackAwareGoal.java

```diff
@@ -0,0 +1,442 @@
+ *
+ */
+
+package com.linkedin.kafka.cruisecontrol.analyzer.kafkaassigner;
+
+import com.linkedin.kafka.cruisecontrol.analyzer.BalancingProposal;
+import com.linkedin.kafka.cruisecontrol.analyzer.goals.Goal;
+import com.linkedin.kafka.cruisecontrol.common.BalancingAction;
+import com.linkedin.kafka.cruisecontrol.exception.AnalysisInputException;
+import com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException;
+import com.linkedin.kafka.cruisecontrol.exception.ModelInputException;
+import com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException;
+import com.linkedin.kafka.cruisecontrol.model.Broker;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModelStats;
+import com.linkedin.kafka.cruisecontrol.model.Partition;
+import com.linkedin.kafka.cruisecontrol.model.Replica;
+import com.linkedin.kafka.cruisecontrol.monitor.ModelCompletenessRequirements;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class KafkaAssignerEvenRackAwareGoal implements Goal {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaAssignerEvenRackAwareGoal.class);
+  private Map<String, List<Partition>> _partitionsByTopic;
+  private Map<Integer, SortedSet<BrokerReplicaCount>> _healthyBrokerReplicaCountByPosition;
+
+  public KafkaAssignerEvenRackAwareGoal() {
+    _partitionsByTopic = null;
+    _healthyBrokerReplicaCountByPosition = new HashMap<>();
+  }
+
+  /**
+   * Optimize the given cluster model as needed for this goal.
+   *
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Goals that have already been optimized. These goals cannot be violated.
+   * @param excludedTopics The topics that should be excluded from the optimization proposal.
+   * @return true if the goal is met after the optimization, throws an exceptions if the goal is not met.
+   */
+  @Override
+  public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set<String> excludedTopics)
+      throws KafkaCruiseControlException {
+    LOG.debug("Starting {} with excluded topics = {}", name(), excludedTopics);
+
+    if (!optimizedGoals.isEmpty()) {
+      throw new IllegalArgumentException(String.format("Goals %s cannot be optimized before %s.", optimizedGoals, name()));
+    }
+
+    ensureRackAwareSatisfiable(clusterModel, excludedTopics);
+
+    _partitionsByTopic = clusterModel.getPartitionsByTopic();
+    int maxReplicationFactor = clusterModel.maxReplicationFactor();
+    Map<String, Integer> replicationFactorByTopic = clusterModel.replicationFactorByTopic();
+
+    for (int i = 0; i < maxReplicationFactor; i++) {
```

Comment:
Should the BrokerReplicaCount initialize with number of replicas of the excluded topics?